CI: concurrent map read and map write in ipmasq tests

```
coverage: 2.1% of statements in ./...
ok  	github.com/cilium/cilium/pkg/ipcache/fake	0.063s	coverage: 2.1% of statements in ./...
=== RUN   Test
fatal error: concurrent map read and map write
goroutine 33 [running]:
runtime.throw(0xc67924, 0x21)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/runtime/panic.go:1116 +0x72 fp=0xc000574bb0 sp=0xc000574b80 pc=0x4370d2
runtime.mapaccess2_faststr(0xb78a00, 0xc00009e810, 0xc00003a320, 0xa, 0xc000574fb8, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/runtime/map_faststr.go:116 +0x47c fp=0xc000574c20 sp=0xc000574bb0 pc=0x4155bc
github.com/cilium/cilium/pkg/ipmasq.(*ipMasqMapMock).Update(0xc0001241c0, 0xc00003a2bc, 0x4, 0x4, 0xc00003a2b8, 0x4, 0x4, 0x0, 0x0)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/ipmasq/ipmasq_test.go:42 +0x6a fp=0xc000574c88 sp=0xc000574c20 pc=0xabc63a
github.com/cilium/cilium/pkg/ipmasq.(*IPMasqAgent).Update(0xc00009e870, 0x22cf4b36, 0xc630b7c75d)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/ipmasq/ipmasq.go:118 +0x324 fp=0xc0005750d8 sp=0xc000574c88 pc=0xabb8a4
github.com/cilium/cilium/pkg/ipmasq.start.func1(0xd71440, 0xc00014ea80, 0x12c7560, 0x0)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/ipmasq/ipmasq.go:102 +0x30 fp=0xc000575100 sp=0xc0005750d8 pc=0xabdb50
github.com/cilium/cilium/pkg/controller.(*Controller).runController(0xc0001c7300)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/controller/controller.go:205 +0xa88 fp=0xc000575fd8 sp=0xc000575100 pc=0xa53ec8
runtime.goexit()
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc000575fe0 sp=0xc000575fd8 pc=0x469241
created by github.com/cilium/cilium/pkg/controller.(*Manager).updateController
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/controller/manager.go:120 +0xb69
goroutine 1 [chan receive]:
testing.(*T).Run(0xc000360d80, 0xc50ee1, 0x4, 0xc8a708, 0x491a76)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:1043 +0x37e
testing.runTests.func1(0xc000360c60)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:1284 +0x78
testing.tRunner(0xc000360c60, 0xc0004f5dd8)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:991 +0xdc
testing.runTests(0xc0004fbf00, 0x12ae940, 0x1, 0x1, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:1282 +0x2a7
testing.(*M).Run(0xc0002b8880, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:1199 +0x15f
main.main()
	_testmain.go:878 +0x1d4
goroutine 22 [chan receive]:
k8s.io/klog.(*loggingT).flushDaemon(0x12c7920)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/klog/klog.go:1010 +0x8b
created by k8s.io/klog.init.0
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/k8s.io/klog/klog.go:411 +0xd6
goroutine 23 [select]:
io.(*pipe).Read(0xc0000889c0, 0xc000562000, 0x1000, 0x1000, 0xb34ba0, 0x1, 0xc000562000)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/io/pipe.go:57 +0xe7
io.(*PipeReader).Read(0xc0001240e8, 0xc000562000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/io/pipe.go:134 +0x4c
bufio.(*Scanner).Scan(0xc000073f38, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/bufio/scan.go:213 +0xa4
github.com/sirupsen/logrus.(*Entry).writerScanner(0xc0001a90a0, 0xc0001240e8, 0xc0004fcef0)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:53 +0xc3
created by github.com/sirupsen/logrus.(*Entry).WriterLevel
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:45 +0x1a5
goroutine 24 [select]:
io.(*pipe).Read(0xc000088a20, 0xc000341000, 0x1000, 0x1000, 0xb34ba0, 0x1, 0xc000341000)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/io/pipe.go:57 +0xe7
io.(*PipeReader).Read(0xc0001240f8, 0xc000341000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/io/pipe.go:134 +0x4c
bufio.(*Scanner).Scan(0xc000071f38, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/bufio/scan.go:213 +0xa4
github.com/sirupsen/logrus.(*Entry).writerScanner(0xc0001a90a0, 0xc0001240f8, 0xc0004fcf10)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:53 +0xc3
created by github.com/sirupsen/logrus.(*Entry).WriterLevel
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:45 +0x1a5
goroutine 25 [select]:
io.(*pipe).Read(0xc000088a80, 0xc000563000, 0x1000, 0x1000, 0xb34ba0, 0x1, 0xc000563000)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/io/pipe.go:57 +0xe7
io.(*PipeReader).Read(0xc000124108, 0xc000563000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/io/pipe.go:134 +0x4c
bufio.(*Scanner).Scan(0xc000074f38, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/bufio/scan.go:213 +0xa4
github.com/sirupsen/logrus.(*Entry).writerScanner(0xc0001a90a0, 0xc000124108, 0xc0004fcf30)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:53 +0xc3
created by github.com/sirupsen/logrus.(*Entry).WriterLevel
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:45 +0x1a5
goroutine 26 [select]:
io.(*pipe).Read(0xc000088ae0, 0xc000564000, 0x1000, 0x1000, 0xb34ba0, 0x1, 0xc000564000)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/io/pipe.go:57 +0xe7
io.(*PipeReader).Read(0xc000124118, 0xc000564000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/io/pipe.go:134 +0x4c
bufio.(*Scanner).Scan(0xc000075f38, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/bufio/scan.go:213 +0xa4
github.com/sirupsen/logrus.(*Entry).writerScanner(0xc0001a90a0, 0xc000124118, 0xc0004fcf50)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:53 +0xc3
created by github.com/sirupsen/logrus.(*Entry).WriterLevel
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/github.com/sirupsen/logrus/writer.go:45 +0x1a5
goroutine 29 [chan receive]:
gopkg.in/check%2ev1.(*suiteRunner).runTest(...)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:819
gopkg.in/check%2ev1.(*suiteRunner).run(0xc0002b8980, 0xc0004fbea0)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:624 +0x11d
gopkg.in/check%2ev1.Run(0xbaafa0, 0xc0004fbea0, 0xc000062f18, 0xd0)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:92 +0x4d
gopkg.in/check%2ev1.RunAll(0xc0005fff18, 0x0)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:84 +0x96
gopkg.in/check%2ev1.TestingT(0xc000360d80)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/run.go:72 +0x395
github.com/cilium/cilium/pkg/ipmasq.Test(0xc000360d80)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/ipmasq/ipmasq_test.go:33 +0x2b
testing.tRunner(0xc000360d80, 0xc8a708)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/testing/testing.go:1042 +0x357
goroutine 30 [select]:
gopkg.in/check%2ev1.(*resultTracker)._loopRoutine(0xc0004f8240)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:470 +0xc4
created by gopkg.in/check%2ev1.(*resultTracker).start
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:450 +0x3f
goroutine 31 [sleep]:
time.Sleep(0x11e1a300)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/runtime/time.go:188 +0xba
github.com/cilium/cilium/pkg/ipmasq.(*IPMasqTestSuite).TestRestore(0xc0004fbea0, 0xc0005921e0)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/ipmasq/ipmasq_test.go:136 +0x34f
reflect.Value.call(0xbaafa0, 0xc0004fbea0, 0xa13, 0xc50dd5, 0x4, 0xc00006ef08, 0x1, 0x1, 0x12f4480, 0xc00006ee48, ...)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/reflect/value.go:460 +0x8ab
reflect.Value.Call(0xbaafa0, 0xc0004fbea0, 0xa13, 0xc00006ef08, 0x1, 0x1, 0xc0005922d0, 0x0, 0x0)
	/home/travis/.gimme/versions/go1.14.2.linux.amd64/src/reflect/value.go:321 +0xb4
gopkg.in/check%2ev1.(*suiteRunner).forkTest.func1(0xc0005921e0)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:781 +0x628
gopkg.in/check%2ev1.(*suiteRunner).forkCall.func1(0xc0002b8980, 0xc0005921e0, 0xc00000e060)
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:675 +0x98
created by gopkg.in/check%2ev1.(*suiteRunner).forkCall
	/home/travis/gopath/src/github.com/cilium/cilium/vendor/gopkg.in/check.v1/check.go:672 +0x1fb
goroutine 34 [select]:
github.com/cilium/cilium/pkg/controller.(*Controller).runController(0xc0001c7400)
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/controller/controller.go:262 +0x1c7
created by github.com/cilium/cilium/pkg/controller.(*Manager).updateController
	/home/travis/gopath/src/github.com/cilium/cilium/pkg/controller/manager.go:120 +0xb69
FAIL	github.com/cilium/cilium/pkg/ipmasq	0.072s
FAIL
Makefile:145: recipe for target 'unit-tests' failed
make: *** [unit-tests] Error 1
The command "./.travis/build.sh" exited with 2.

```

--- a/Documentation/cmdref/cilium-agent.md
+++ b/Documentation/cmdref/cilium-agent.md
@@ -108,7 +108,6 @@ cilium-agent [flags]
       --install-iptables-rules                        Install base iptables rules for cilium to mainly interact with kube-proxy (and masquerading) (default true)
       --ip-allocation-timeout duration                Time after which an incomplete CIDR allocation is considered failed (default 2m0s)
       --ip-masq-agent-config-path string              ip-masq-agent configuration file path (default "/etc/config/ip-masq-agent")
-      --ip-masq-agent-sync-period duration            ip-masq-agent configuration file synchronization period (default 1m0s)
       --ipam string                                   Backend to use for IPAM (default "hostscope-legacy")
       --ipsec-key-file string                         Path to IPSec key file
       --ipv4-node string                              IPv4 address of node (default "auto")

--- a/daemon/cmd/daemon_main.go
+++ b/daemon/cmd/daemon_main.go
@@ -547,9 +547,6 @@ func init() {
 	flags.String(option.IPMasqAgentConfigPath, "/etc/config/ip-masq-agent", "ip-masq-agent configuration file path")
 	option.BindEnv(option.IPMasqAgentConfigPath)
 
-	flags.Duration(option.IPMasqAgentSyncPeriod, 60*time.Second, "ip-masq-agent configuration file synchronization period")
-	option.BindEnv(option.IPMasqAgentSyncPeriod)
-
 	flags.Bool(option.InstallIptRules, true, "Install base iptables rules for cilium to mainly interact with kube-proxy (and masquerading)")
 	option.BindEnv(option.InstallIptRules)
 
@@ -1295,8 +1292,11 @@ func runDaemon() {
 	}
 
 	if option.Config.EnableIPMasqAgent {
-		ipmasq.Start(option.Config.IPMasqAgentConfigPath,
-			option.Config.IPMasqAgentSyncPeriod)
+		ipmasqAgent, err := ipmasq.NewIPMasqAgent(option.Config.IPMasqAgentConfigPath)
+		if err != nil {
+			log.WithError(err).Fatal("Failed to create ip-masq-agent")
+		}
+		ipmasqAgent.Start()
 	}
 
 	if !option.Config.DryMode {

--- a/pkg/ipmasq/ipmasq.go
+++ b/pkg/ipmasq/ipmasq.go
@@ -16,6 +16,7 @@ package ipmasq
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 	"io/ioutil"
 	"net"
@@ -23,7 +24,7 @@ import (
 	"strings"
 	"time"
 
-	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/util/yaml"
 
 	"github.com/cilium/cilium/pkg/controller"
 	"github.com/cilium/cilium/pkg/logging"
@@ -38,10 +39,11 @@ var (
 // ipnet is a wrapper type for net.IPNet to enable de-serialization of CIDRs
 type Ipnet net.IPNet
 
-func (c *Ipnet) UnmarshalYAML(unmarshal func(interface{}) error) error {
-	var str string
-	if err := unmarshal(&str); err != nil {
-		return err
+func (c *Ipnet) UnmarshalJSON(json []byte) error {
+	str := string(json)
+
+	if json[0] != '"' {
+		return fmt.Errorf("Invalid CIDR: %s", str)
 	}
 
 	ip, n, err := net.ParseCIDR(strings.Trim(str, `"`))
@@ -54,12 +56,11 @@ func (c *Ipnet) UnmarshalYAML(unmarshal func(interface{}) error) error {
 
 	*c = Ipnet(*n)
 	return nil
-
 }
 
 // config represents the ip-masq-agent configuration file encoded as YAML
 type config struct {
-	NonMasqCIDRs []Ipnet `yaml:"nonMasqueradeCIDRs"`
+	NonMasqCIDRs []Ipnet `json:"nonMasqueradeCIDRs"`
 }
 
 // IPMasqMap is an interface describing methods for manipulating an ipmasq map
@@ -146,8 +147,13 @@ func (a *IPMasqAgent) readConfig() error {
 		return fmt.Errorf("Failed to read %s: %s", a.configPath, err)
 	}
 
-	if err := yaml.Unmarshal(raw, &cfg); err != nil {
-		return fmt.Errorf("Failed to de-serialize yaml: %s", err)
+	jsonStr, err := yaml.ToJSON(raw)
+	if err != nil {
+		return fmt.Errorf("Failed to convert to json: %s", err)
+	}
+
+	if err := json.Unmarshal([]byte(jsonStr), &cfg); err != nil {
+		return fmt.Errorf("Failed to de-serialize json: %s", err)
 	}
 
 	nonMasqCIDRs := map[string]net.IPNet{}

--- a/pkg/ipmasq/ipmasq.go
+++ b/pkg/ipmasq/ipmasq.go
@@ -15,18 +15,17 @@
 package ipmasq
 
 import (
-	"context"
 	"encoding/json"
 	"fmt"
 	"io/ioutil"
 	"net"
 	"os"
+	"path/filepath"
 	"strings"
-	"time"
 
+	"github.com/fsnotify/fsnotify"
 	"k8s.io/apimachinery/pkg/util/yaml"
 
-	"github.com/cilium/cilium/pkg/controller"
 	"github.com/cilium/cilium/pkg/logging"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/maps/ipmasq"
@@ -76,35 +75,86 @@ type IPMasqAgent struct {
 	nonMasqCIDRsFromConfig map[string]net.IPNet
 	nonMasqCIDRsInMap      map[string]net.IPNet
 	ipMasqMap              IPMasqMap
+	watcher                *fsnotify.Watcher
+	stop                   chan struct{}
+	handlerFinished        chan struct{}
 }
 
-// Start starts the "ip-masq-agent" controller which is used to sync the ipmasq
-// BPF maps.
-func Start(configPath string, syncPeriod time.Duration) {
-	start(configPath, syncPeriod, &ipmasq.IPMasqBPFMap{}, controller.NewManager())
+func NewIPMasqAgent(configPath string) (*IPMasqAgent, error) {
+	return newIPMasqAgent(configPath, &ipmasq.IPMasqBPFMap{})
 }
 
-func start(configPath string, syncPeriod time.Duration,
-	ipMasqMap IPMasqMap, manager *controller.Manager) {
+func newIPMasqAgent(configPath string, ipMasqMap IPMasqMap) (*IPMasqAgent, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create fsnotify watcher: %s", err)
+	}
+
+	configDir := filepath.Dir(configPath)
+	// The directory of the config should exist at this time, otherwise
+	// the watcher will fail to add
+	if err := watcher.Add(configDir); err != nil {
+		watcher.Close()
+		return nil, fmt.Errorf("Failed to add %q dir to fsnotify watcher: %s", configDir, err)
+	}
 
 	a := &IPMasqAgent{
 		configPath:        configPath,
 		nonMasqCIDRsInMap: map[string]net.IPNet{},
 		ipMasqMap:         ipMasqMap,
+		watcher:           watcher,
 	}
 
+	return a, nil
+}
+
+// Start starts the ip-masq-agent goroutine which tracks the config file and
+// updates the BPF map accordingly.
+func (a *IPMasqAgent) Start() {
 	if err := a.restore(); err != nil {
-		log.WithError(err).Warn("ip-masq-agent failed to restore")
+		log.WithError(err).Warn("Failed to restore")
+	}
+	if err := a.Update(); err != nil {
+		log.WithError(err).Warn("Failed to update")
 	}
 
-	manager.UpdateController("ip-masq-agent",
-		controller.ControllerParams{
-			DoFunc: func(ctx context.Context) error {
-				return a.Update()
-			},
-			RunInterval: syncPeriod,
-		},
-	)
+	a.stop = make(chan struct{})
+	a.handlerFinished = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case event := <-a.watcher.Events:
+				log.Debugf("Received fsnotify event: %+v", event)
+
+				if event.Name != a.configPath {
+					continue
+				}
+
+				switch event.Op {
+				case fsnotify.Create, fsnotify.Write, fsnotify.Chmod, fsnotify.Remove, fsnotify.Rename:
+					if err := a.Update(); err != nil {
+						log.WithError(err).Warn("Failed to update")
+					}
+				default:
+					log.Warnf("Watcher received unknown event: %s. Ignoring.", event)
+				}
+			case err := <-a.watcher.Errors:
+				log.WithError(err).Warn("Watcher received an error")
+			case <-a.stop:
+				log.Info("Stopping ip-masq-agent")
+				close(a.handlerFinished)
+				return
+			}
+		}
+	}()
+}
+
+// Stop stops the ip-masq-agent goroutine and the watcher.
+func (a *IPMasqAgent) Stop() {
+	close(a.stop)
+	<-a.handlerFinished
+	a.watcher.Close()
 }
 
 // Update updates the ipmasq BPF map entries with ones from the config file.
@@ -152,7 +202,7 @@ func (a *IPMasqAgent) readConfig() error {
 		return fmt.Errorf("Failed to convert to json: %s", err)
 	}
 
-	if err := json.Unmarshal([]byte(jsonStr), &cfg); err != nil {
+	if err := json.Unmarshal(jsonStr, &cfg); err != nil {
 		return fmt.Errorf("Failed to de-serialize json: %s", err)
 	}
 

--- a/pkg/ipmasq/ipmasq_test.go
+++ b/pkg/ipmasq/ipmasq_test.go
@@ -26,7 +26,6 @@ import (
 
 	"gopkg.in/check.v1"
 
-	"github.com/cilium/cilium/pkg/controller"
 	"github.com/cilium/cilium/pkg/lock"
 )
 
@@ -90,28 +89,28 @@ func (m *ipMasqMapMock) dumpToSet() map[string]struct{} {
 }
 
 type IPMasqTestSuite struct {
-	ipMasqMap  *ipMasqMapMock
-	manager    *controller.Manager
-	configFile *os.File
+	ipMasqMap   *ipMasqMapMock
+	ipMasqAgent *IPMasqAgent
+	configFile  *os.File
 }
 
 var _ = check.Suite(&IPMasqTestSuite{})
 
 func (i *IPMasqTestSuite) SetUpTest(c *check.C) {
 	i.ipMasqMap = &ipMasqMapMock{cidrs: map[string]net.IPNet{}}
-	i.manager = controller.NewManager()
 
 	configFile, err := ioutil.TempFile("", "ipmasq-test")
 	c.Assert(err, check.IsNil)
 	i.configFile = configFile
 
-	start(configFile.Name(), 100*time.Millisecond, i.ipMasqMap, i.manager)
+	agent, err := newIPMasqAgent(configFile.Name(), i.ipMasqMap)
+	c.Assert(err, check.IsNil)
+	i.ipMasqAgent = agent
+	i.ipMasqAgent.Start()
 }
 
 func (i *IPMasqTestSuite) TearDownTest(c *check.C) {
-	err := i.manager.RemoveController("ip-masq-agent")
-	c.Assert(err, check.IsNil)
-
+	i.ipMasqAgent.Stop()
 	os.Remove(i.configFile.Name())
 }
 
@@ -158,28 +157,47 @@ func (i *IPMasqTestSuite) TestUpdate(c *check.C) {
 	// Delete file, should remove the CIDRs
 	err = os.Remove(i.configFile.Name())
 	c.Assert(err, check.IsNil)
+	err = i.configFile.Close()
+	c.Assert(err, check.IsNil)
 	time.Sleep(300 * time.Millisecond)
 	ipnets = i.ipMasqMap.dumpToSet()
 	c.Assert(len(ipnets), check.Equals, 0)
 }
 
 func (i *IPMasqTestSuite) TestRestore(c *check.C) {
-	err := i.manager.RemoveController("ip-masq-agent")
-	c.Assert(err, check.IsNil)
+	// Stop ip-masq-agent goroutine (we can't use i.ipMasqMap.Stop(), as it stops
+	// the watcher)
+	close(i.ipMasqAgent.stop)
 
 	_, cidr, _ := net.ParseCIDR("3.3.3.0/24")
 	i.ipMasqMap.cidrs[cidr.String()] = *cidr
 	_, cidr, _ = net.ParseCIDR("4.4.0.0/16")
 	i.ipMasqMap.cidrs[cidr.String()] = *cidr
 
-	_, err = i.configFile.WriteString("nonMasqueradeCIDRs:\n- 4.4.0.0/16")
+	_, err := i.configFile.WriteString("nonMasqueradeCIDRs:\n- 4.4.0.0/16")
 	c.Assert(err, check.IsNil)
 
-	start(i.configFile.Name(), 100*time.Millisecond, i.ipMasqMap, i.manager)
+	i.ipMasqAgent.Start()
 	time.Sleep(300 * time.Millisecond)
 
 	ipnets := i.ipMasqMap.dumpToSet()
 	c.Assert(len(ipnets), check.Equals, 1)
 	_, ok := ipnets["4.4.0.0/16"]
+	c.Assert(ok, check.Equals, true)
+
+	// Now stop the goroutine, and also remove the maps. It should bootstrap from
+	// the config
+	close(i.ipMasqAgent.stop)
+	i.ipMasqMap = &ipMasqMapMock{cidrs: map[string]net.IPNet{}}
+	i.ipMasqAgent.ipMasqMap = i.ipMasqMap
+	_, err = i.configFile.Seek(0, 0)
+	c.Assert(err, check.IsNil)
+	_, err = i.configFile.WriteString("nonMasqueradeCIDRs:\n- 3.3.0.0/16")
+	c.Assert(err, check.IsNil)
+	i.ipMasqAgent.Start()
+
+	ipnets = i.ipMasqMap.dumpToSet()
+	c.Assert(len(ipnets), check.Equals, 1)
+	_, ok = ipnets["3.3.0.0/16"]
 	c.Assert(ok, check.Equals, true)
 }

--- a/pkg/ipmasq/ipmasq_test.go
+++ b/pkg/ipmasq/ipmasq_test.go
@@ -113,6 +113,19 @@ func (i *IPMasqTestSuite) TestUpdate(c *check.C) {
 	_, ok = i.ipMasqMap.cidrs["2.2.0.0/16"]
 	c.Assert(ok, check.Equals, true)
 
+	// Write new config in JSON
+	_, err = i.configFile.Seek(0, 0)
+	c.Assert(err, check.IsNil)
+	_, err = i.configFile.WriteString(`{"nonMasqueradeCIDRs": ["8.8.0.0/16", "1.1.2.3/16"]}`)
+	c.Assert(err, check.IsNil)
+	time.Sleep(300 * time.Millisecond)
+
+	c.Assert(len(i.ipMasqMap.cidrs), check.Equals, 2)
+	_, ok = i.ipMasqMap.cidrs["8.8.0.0/16"]
+	c.Assert(ok, check.Equals, true)
+	_, ok = i.ipMasqMap.cidrs["1.1.0.0/16"]
+	c.Assert(ok, check.Equals, true)
+
 	// Delete file, should remove the CIDRs
 	err = os.Remove(i.configFile.Name())
 	c.Assert(err, check.IsNil)

--- a/pkg/ipmasq/ipmasq_test.go
+++ b/pkg/ipmasq/ipmasq_test.go
@@ -27,6 +27,7 @@ import (
 	"gopkg.in/check.v1"
 
 	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/lock"
 )
 
 func Test(t *testing.T) {
@@ -34,33 +35,58 @@ func Test(t *testing.T) {
 }
 
 type ipMasqMapMock struct {
+	lock.RWMutex
 	cidrs map[string]net.IPNet
 }
 
 func (m *ipMasqMapMock) Update(cidr net.IPNet) error {
+	m.Lock()
+	defer m.Unlock()
+
 	cidrStr := cidr.String()
 	if _, ok := m.cidrs[cidrStr]; ok {
 		return fmt.Errorf("CIDR already exists: %s", cidrStr)
 	}
 	m.cidrs[cidrStr] = cidr
+
 	return nil
 }
 
 func (m *ipMasqMapMock) Delete(cidr net.IPNet) error {
+	m.Lock()
+	defer m.Unlock()
+
 	cidrStr := cidr.String()
 	if _, ok := m.cidrs[cidrStr]; !ok {
 		return fmt.Errorf("CIDR not found: %s", cidrStr)
 	}
 	delete(m.cidrs, cidrStr)
+
 	return nil
 }
 
 func (m *ipMasqMapMock) Dump() ([]net.IPNet, error) {
+	m.RLock()
+	defer m.RUnlock()
+
 	cidrs := make([]net.IPNet, 0, len(m.cidrs))
 	for _, cidr := range m.cidrs {
 		cidrs = append(cidrs, cidr)
 	}
+
 	return cidrs, nil
+}
+
+func (m *ipMasqMapMock) dumpToSet() map[string]struct{} {
+	m.RLock()
+	defer m.RUnlock()
+
+	cidrs := make(map[string]struct{}, len(m.cidrs))
+	for cidrStr := range m.cidrs {
+		cidrs[cidrStr] = struct{}{}
+	}
+
+	return cidrs
 }
 
 type IPMasqTestSuite struct {
@@ -94,10 +120,11 @@ func (i *IPMasqTestSuite) TestUpdate(c *check.C) {
 	c.Assert(err, check.IsNil)
 	time.Sleep(300 * time.Millisecond)
 
-	c.Assert(len(i.ipMasqMap.cidrs), check.Equals, 2)
-	_, ok := i.ipMasqMap.cidrs["1.1.1.1/32"]
+	ipnets := i.ipMasqMap.dumpToSet()
+	c.Assert(len(ipnets), check.Equals, 2)
+	_, ok := ipnets["1.1.1.1/32"]
 	c.Assert(ok, check.Equals, true)
-	_, ok = i.ipMasqMap.cidrs["2.2.0.0/16"]
+	_, ok = ipnets["2.2.0.0/16"]
 	c.Assert(ok, check.Equals, true)
 
 	// Write new config
@@ -107,10 +134,11 @@ func (i *IPMasqTestSuite) TestUpdate(c *check.C) {
 	c.Assert(err, check.IsNil)
 	time.Sleep(300 * time.Millisecond)
 
-	c.Assert(len(i.ipMasqMap.cidrs), check.Equals, 2)
-	_, ok = i.ipMasqMap.cidrs["8.8.0.0/16"]
+	ipnets = i.ipMasqMap.dumpToSet()
+	c.Assert(len(ipnets), check.Equals, 2)
+	_, ok = ipnets["8.8.0.0/16"]
 	c.Assert(ok, check.Equals, true)
-	_, ok = i.ipMasqMap.cidrs["2.2.0.0/16"]
+	_, ok = ipnets["2.2.0.0/16"]
 	c.Assert(ok, check.Equals, true)
 
 	// Write new config in JSON
@@ -120,17 +148,19 @@ func (i *IPMasqTestSuite) TestUpdate(c *check.C) {
 	c.Assert(err, check.IsNil)
 	time.Sleep(300 * time.Millisecond)
 
-	c.Assert(len(i.ipMasqMap.cidrs), check.Equals, 2)
-	_, ok = i.ipMasqMap.cidrs["8.8.0.0/16"]
+	ipnets = i.ipMasqMap.dumpToSet()
+	c.Assert(len(ipnets), check.Equals, 2)
+	_, ok = ipnets["8.8.0.0/16"]
 	c.Assert(ok, check.Equals, true)
-	_, ok = i.ipMasqMap.cidrs["1.1.0.0/16"]
+	_, ok = ipnets["1.1.0.0/16"]
 	c.Assert(ok, check.Equals, true)
 
 	// Delete file, should remove the CIDRs
 	err = os.Remove(i.configFile.Name())
 	c.Assert(err, check.IsNil)
 	time.Sleep(300 * time.Millisecond)
-	c.Assert(len(i.ipMasqMap.cidrs), check.Equals, 0)
+	ipnets = i.ipMasqMap.dumpToSet()
+	c.Assert(len(ipnets), check.Equals, 0)
 }
 
 func (i *IPMasqTestSuite) TestRestore(c *check.C) {
@@ -148,7 +178,8 @@ func (i *IPMasqTestSuite) TestRestore(c *check.C) {
 	start(i.configFile.Name(), 100*time.Millisecond, i.ipMasqMap, i.manager)
 	time.Sleep(300 * time.Millisecond)
 
-	c.Assert(len(i.ipMasqMap.cidrs), check.Equals, 1)
-	_, ok := i.ipMasqMap.cidrs["4.4.0.0/16"]
+	ipnets := i.ipMasqMap.dumpToSet()
+	c.Assert(len(ipnets), check.Equals, 1)
+	_, ok := ipnets["4.4.0.0/16"]
 	c.Assert(ok, check.Equals, true)
 }

--- a/pkg/option/config.go
+++ b/pkg/option/config.go
@@ -276,9 +276,6 @@ const (
 	// IPMasqAgentConfigPath is the configuration file path
 	IPMasqAgentConfigPath = "ip-masq-agent-config-path"
 
-	// IPMasqAgentSyncPeriod is the configuration file synchronization period
-	IPMasqAgentSyncPeriod = "ip-masq-agent-sync-period"
-
 	// InstallIptRules sets whether Cilium should install any iptables in general
 	InstallIptRules = "install-iptables-rules"
 
@@ -1454,7 +1451,6 @@ type DaemonConfig struct {
 	EnableBPFMasquerade    bool
 	EnableIPMasqAgent      bool
 	IPMasqAgentConfigPath  string
-	IPMasqAgentSyncPeriod  time.Duration
 	InstallIptRules        bool
 	MonitorAggregation     string
 	PreAllocateMaps        bool
@@ -2231,7 +2227,6 @@ func (c *DaemonConfig) Populate() {
 	c.EnableBPFMasquerade = viper.GetBool(EnableBPFMasquerade)
 	c.EnableIPMasqAgent = viper.GetBool(EnableIPMasqAgent)
 	c.IPMasqAgentConfigPath = viper.GetString(IPMasqAgentConfigPath)
-	c.IPMasqAgentSyncPeriod = viper.GetDuration(IPMasqAgentSyncPeriod)
 	c.InstallIptRules = viper.GetBool(InstallIptRules)
 	c.IPSecKeyFile = viper.GetString(IPSecKeyFileName)
 	c.ModePreFilter = viper.GetString(PrefilterMode)

